File: src/main/java/com/typelead/intellij/utils/parser/SimplePsiParser.java
package com.typelead.intellij.utils.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

/**
 * Simple parser which returns a flat tree tokens from the lexer.
 * Useful for providing a stub parser for languages which don't need a parse tree.
 */
public class SimplePsiParser implements PsiParser {
    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker marker = builder.mark();
        while (!builder.eof()) {
            builder.advanceLexer();
        }
        marker.done(root);
        return builder.getTreeBuilt();
    }
}

File: src/main/eta/FFI/Com/TypeLead/IntelliJ/Utils/Parser/SimplePsiParser.hs
module FFI.Com.TypeLead.IntelliJ.Utils.Parser.SimplePsiParser where

import P

import FFI.Com.IntelliJ.Lang.PsiParser

data {-# CLASS "com.typelead.intellij.utils.parser.SimplePsiParser" #-}
  SimplePsiParser = SimplePsiParser (Object# SimplePsiParser)
  deriving Class

type instance Inherits SimplePsiParser = '[PsiParser]

foreign import java unsafe "@new" newSimplePsiParser :: Java a SimplePsiParser
